packages/cubejs-api-gateway/src/helpers/sqlQueryHandler.ts

```diff
@@ -0,0 +1,103 @@
+import { UserError } from '../UserError';
+
+export const enum SQlRunnerQueryType {
+  Select = 'select',
+  With = 'with',
+  Create = 'create',
+  Alter = 'alter',
+  Drop = 'drop',
+  Truncate = 'truncate',
+  Insert = 'insert',
+  Update = 'update',
+  Delete = 'delete',
+  Explain = 'explain',
+}
+
+// Determine SQL query type and add LIMIT clause if needed
+export const shouldAddLimit = (sql: string): Boolean => {
+  const ddlRegex = /^(CREATE|ALTER|DROP|TRUNCATE)\b/i;
+  const dmlRegex = /^(INSERT|UPDATE|DELETE)\b/i;
+  const explainRegexp = /^(EXPLAIN)\b/i;
+  const selectRegex = /^(SELECT|WITH)\b/i;
+
+  if (ddlRegex.test(sql) || dmlRegex.test(sql) || explainRegexp.test(sql)) {
+    return false;
+  } else if (selectRegex.test(sql)) {
+    return true;
+  }
+
+  throw new UserError('Invalid SQL query');
+};
+
+// Determine particular SQL query type
+const getStatementType = (sql: string): string | null => {
+  if (/^(CREATE)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Create;
+  }
+
+  if (/^(ALTER)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Alter;
+  }
+
+  if (/^(DROP)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Drop;
+  }
+
+  if (/^(TRUNCATE)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Truncate;
+  }
+
+  if (/^(INSERT)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Insert;
+  }
+
+  if (/^(UPDATE)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Update;
+  }
+
+  if (/^(DELETE)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Delete;
+  }
+
+  if (/^(SELECT)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Select;
+  }
+
+  if (/^(WITH)\b/i.test(sql)) {
+    return SQlRunnerQueryType.With;
+  }
+
+  if (/^(EXPLAIN)\b/i.test(sql)) {
+    return SQlRunnerQueryType.Explain;
+  }
+
+  return null;
+};
+
+// check query string for scope permissions
+// an example scope with permissions: ['sql-runner-permissions:create,update,delete,select']
+export const isQueryHasPermission = (
+  scope: string[],
+  query: string
+): Boolean => {
+  const permissionsString = scope.find((s: string) => s.startsWith('sql-runner-permissions:'));
+
+  if (!permissionsString) {
+    return true;
+  }
+
+  const [, p] = permissionsString.split(':');
+  if (!p.length) {
+    return false;
+  }
+
+  const permissions = p.split(',');
+
+  const statementType = getStatementType(query);
+
+  if (statementType !== null && permissions.includes(statementType)) {
+    return true;
+  }
+
+  return false;
+};
```

Comment:
Please consider reviewing the code for better readability.
Also, please change the name of the `isQueryHasPermission` to smth like `isQueryAllowed`